### ISSUE TITLE
App crash: Cannot handle event while being in inappropriate state

### ISSUE THREAD

**Issue by xzhorikx** [GetStream/stream-chat-android#1405]
**Describe the bug**
App is crashing  *sometimes* when 
* Trying to connect the anonymous user to the client - `chatClient.connectAnonymousUser().execute()`
* Trying to disconnect - `chatClient.disconnect()`
Application is crashing with following exception
```
    java.lang.IllegalStateException: Cannot handle event io.getstream.chat.android.client.clientstate.ClientStateService$ClientStateEvent$DisconnectedEvent@9ea6031 while being in inappropriate state io.getstream.chat.android.client.clientstate.ClientState$Anonymous$Pending$WithoutToken@16139d7
```
This bug is non-consistent and I'm not sure what is the initial cause. Sometimes all requests are going through just fine, sometimes the exceptions produced by the requests are caught in the catch block, and sometimes the whole application crashes. All of that with the same stacktrace:
```
2021-02-15 18:22:29.301 3898-3898/my.app E/AndroidRuntime: FATAL EXCEPTION: main
    Process: my.app, PID: 1234
    java.lang.IllegalStateException: Cannot handle event io.getstream.chat.android.client.clientstate.ClientStateService$ClientStateEvent$DisconnectedEvent@9ea6031 while being in inappropriate state io.getstream.chat.android.client.clientstate.ClientState$Anonymous$Pending$WithoutToken@16139d7
        at io.getstream.chat.android.client.clientstate.ClientStateService.failedToHandleEvent(ClientStateService.kt:126)
        at io.getstream.chat.android.client.clientstate.ClientStateService.access$failedToHandleEvent(ClientStateService.kt:6)
        at io.getstream.chat.android.client.clientstate.ClientStateService$stateMachine$2$1$1.invoke(ClientStateService.kt:35)
        at io.getstream.chat.android.client.clientstate.ClientStateService$stateMachine$2$1$1.invoke(ClientStateService.kt:6)
        at io.getstream.chat.android.core.internal.fsm.FiniteStateMachine$sendEvent$1.invokeSuspend(FiniteStateMachine.kt:56)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:84)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source:1)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source:1)
        at io.getstream.chat.android.core.internal.fsm.FiniteStateMachine.sendEvent(FiniteStateMachine.kt:49)
        at io.getstream.chat.android.client.clientstate.ClientStateService.onDisconnected(ClientStateService.kt:12)
        at io.getstream.chat.android.client.ChatClient$1.onEvent(ChatClient.kt:120)
        at io.getstream.chat.android.client.utils.observable.SubscriptionImpl.onNext(Subscriptions.kt:36)
        at io.getstream.chat.android.client.utils.observable.ChatEventsObservable.onNext(ChatEventsObservable.kt:29)
        at io.getstream.chat.android.client.utils.observable.ChatEventsObservable.access$onNext(ChatEventsObservable.kt:17)
        at io.getstream.chat.android.client.utils.observable.ChatEventsObservable$EventsMapper.onDisconnected(ChatEventsObservable.kt:93)
        at io.getstream.chat.android.client.socket.ChatSocketServiceImpl$state$2$5.invoke(ChatSocketServiceImpl.kt:81)
        at io.getstream.chat.android.client.socket.ChatSocketServiceImpl$state$2$5.invoke(ChatSocketServiceImpl.kt:18)
        at io.getstream.chat.android.client.socket.ChatSocketServiceImpl$callListeners$$inlined$synchronized$lambda$1.run(ChatSocketServiceImpl.kt:202)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:359)
        at android.app.ActivityThread.main(ActivityThread.java:7418)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:935)

```

**SDK version**
- 4.5.4

**To Reproduce**
Initialize the `ChatClient` in the Main Application class
```
import io.getstream.chat.android.client.ChatClient

class MyApp : Application() {
    val GETSTREAM_API_KEY = "your_key"
    lateinit var chat: ChatClient
    override fun onCreate() {
         chat = ChatClient.Builder(GETSTREAM_API_KEY, this).build()
    }
}
```
Make the calls to the `ChatClient` from the main actiivty:
```
import kotlinx.coroutines.*
import io.getstream.chat.android.client.ChatClient

class MainActivity : Activity() {
    override fun onResume() {
        super.onResume()
        GlobalScope.launch(Dispatchers.IO) { 
            val chatClient = ChatClient.instance()
            chatClient.disconnect()
            val result = chatClient
                .connectUser(user, streamChatTokenProvider)
                .execute()    
        }    
    }
}
```

**Expected behavior**
Expected for the anonymous user to be set all the times without the crash.

**Device:**
 - Vendor and model: Xperia 10 plus
 - Android version: 10

**Screenshots**
Nonapplicable

**Comment by JcMinarro** [comment]
Hello @xzhorikx 
Thanks for the report.
We are going to investigate this error

**Comment by zsmb13** [comment]
This has been fixed [4.6.0](https://github.com/GetStream/stream-chat-android/releases/tag/4.6.0), please update and let us know if you have any more questions.
